---
title: "Introduction to single cell RNA-seq analysis"
author: Rene Welch
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{intro_sc_analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  message = FALSE, warning = FALSE, fig.align = "center",
  collapse = TRUE,
  comment = "#>"
)
```

## Workshop description

In exploring and analyzing single cell RNA sequencing data, there are a number of key concepts, such as filtering, scaling, dimensionality reduction, hypothesis testing, clustering and visualization, that need to be understood.

The idea behind this small session is to provide tools to start analyzing a single cell RNA-seq dataset, we are going to focus on how to utilize the R package Seurat to provide a dimensionality reduction visualization. These notes are based on this [tutorial](https://satijalab.org/seurat/articles/pbmc3k_tutorial.html)

### R / Bioconductor package used

* `scRNAseq`
* `Seurat`

## Introduction

There are many steps involved in analyzing a single cell RNA-seq experiment. Usually, we start by aligning the sequence reads into a reference genome to quantify the number of reads mapped to each gene. This results into a table of counts, which we are going to use to perform statistical analysis using R. For single cell RNA-seq data, there are many tools to perform this step, but today we are going to utilize from an already curated dataset, and show how to process it with the R package `Seurat`.

First, let's load all the packages we will need to analyze the data:

```{r, message=FALSE,warning=FALSE}
library(Seurat) ## main package that we are going to use
library(SingleCellExperiment)
library(scRNAseq) 
library(tidyverse) 
library(cowplot)
```

## Getting the data

First, we need to download a dataset. Fortunately, the `scRNAseq` package provides a variety of already curated datasets:

```{r lung-data, message=FALSE, echo=TRUE,include=TRUE, eval=TRUE, warning=FALSE}
lung_data <- ZilionisLungData("mouse")
colnames(lung_data) <- make.unique(colnames(lung_data), sep = "_")
colData(lung_data)[["Barcode"]] <- colnames(lung_data)
logcounts(lung_data) <- as.matrix(log10(1 + counts(lung_data)))
lung_data <- as.Seurat(lung_data, project = "lung")
```

Now, Seurat provides all sorts of useful information on our dataset:

```{r include=TRUE,echo=TRUE,eval=TRUE}
lung_data
```

To access the metadata, i.e. information for each cell / droplet, we do:

```{r include=TRUE,echo=TRUE,eval=TRUE}
head(lung_data@meta.data)
```

This is a curated dataset, thus it already contains the cell labels. At the moment we are going to assume that we don't know that yet.

**Quick note:** This data could be a bit large for some laptops, so in case of trouble running this naalysis, then it could be useful to pick another dataset. To list the datasets provided by the `scRNAseq` package use:

```{r eval = FALSE}
listDatasets()
```

## Quality control

Usually to do a quick quality control, the idea behind this step is that we would like to not have cells in our dataset with an abnormally large / small # of genes (we quantify the # of genes for a cell/droplet as the # of genes with at least 1 read). The reasons for them are:

- Cell doublets or multiplets may exhibit and abnormally high number of genes
- Low-quality cells or empty droplets will often have few genes

First, we add a pair of QC metric to the `meta.data`, and the explore their distribution:

```{r brief-qc, include=TRUE,echo=TRUE,eval=TRUE, out.width = "75%", fig.height=4,fig.width=7}
lung_data[["nCount_RNA"]] <- colSums(lung_data[["RNA"]]@counts)
lung_data[["nFeature_RNA"]] <- apply(lung_data[["RNA"]]@counts, 2, function(x)sum(x > 0))
VlnPlot(lung_data, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
```

Then, we compare those quantities, as expected both metrics are very correlated:

```{r metrics-corr, fig.width=6,fig.height=4, out.width="80%"}
FeatureScatter(lung_data, feature1 = "nCount_RNA", feature2 = "nFeature_RNA",
               cols = "black") +
    geom_smooth(se = FALSE) +
    geom_vline(xintercept = 12e3, linetype = 2, colour = "red") +
    geom_hline(yintercept = 3e3, linetype = 2, colour = "red")
```

We remove the cells with more than 12K reads or more than 3K genes and normalize the data, the default method uses logarithms with a _pseudo-count_ and a scaling factor, which could be good for a first pass. However, many more methods for normalization have been developed, thus it may is possible that there are methods more useful for different datasets.


```{r subset-normalize}
lung_data <- subset(lung_data,
                    subset = nFeature_RNA <= 3e3 & nFeature_RNA <= 12e3)
## writing the default parameters, explicitly
lung_data <- NormalizeData(lung_data,
                           normalization.method = "LogNormalize",
                           scale.factor = 10000)
```

## Identification of most variable genes

The figure shows some association between the average and variances gene expression. The point at the end is that focusing in the most variable genes helps to highlight the signal of single cell datasets. We are going to identify the top 2K most variable genes, but lets keep in mind that this may remove some genes of interest.

```{r variable-features, include=TRUE,echo=TRUE,eval=TRUE,message=FALSE,warning=FALSE, fig.width = 7,fig.height=5,out.width="80%"}

lung_data <- FindVariableFeatures(lung_data, selection.method = "vst",
                                  nfeatures = 2000)
plot1 <- VariableFeaturePlot(lung_data, log = TRUE)
top_genes <- head(VariableFeatures(lung_data), 20)
LabelPoints(plot1, points = top_genes, repel = TRUE)

```

## Principal Components Analysis

The next steps are to scale the data and to compute the principal components of the data. This is necessary because `Seurat` will utilize the components to calculate how do cells relate to each other. These steps are slow when we use all the genes in the data, so it is common to use the list of most variable genes.

```{r pca-analysis}
variable_genes <- VariableFeatures(object = lung_data)
lung_data <- ScaleData(lung_data, features = variable_genes)
lung_data <- RunPCA(lung_data, features = VariableFeatures(object = lung_data))
```

For example, we can visualize top genes that are driving the variance explained by each component:

```{r viz-loadings, fig.width=8,fig.height=5,out.width="90%"}
VizDimLoadings(lung_data, dims = 1:3, reduction = "pca", ncol = 3)
```
Then, the next question to solve is how many components to pick. There are many alternatives to solve this question, but the simplest alternative is to pick them by the Elbow criteria, which is to compare the std. deviation explained by each components with its rank, and pick the previous components before we observe and elbow pattern.

```{r pick-pc-components}
ElbowPlot(lung_data)
```

This is not an exact approach, so we according to the figure above, we would pick the top 2, 7 or 12 components.

## Clustering the cells

This is the most important part of the analysis so far, `Seurat` applies a graph based clustering approach, and to determine the number of clusters it is going to be necessary to tune the two parameters: __the top # of principal components__ and the __resolution__.

- The __top # principal components__ determines how detailed is going to be our approximation when computing the distance between any two cells
- The __resolution__ determines the size of the clusters that the algorithm finds. Smaller values will result in larger clusters or less clusters.

`Seurat` utilizes the Louvain algorithm to cluster the cells in a neighbor graph, and it works by maximizing the "modularity" of the cells network, so it tries to group together the nodes that very similar among them, and to maximize the distance between different clusters / modules.

```{r clustering}
lung_data <- FindNeighbors(lung_data, dims = 1:7)
lung_data <- FindClusters(lung_data, resolution = .3)
## this is to make the cluster colors agree below
nclusters <- nlevels(lung_data@meta.data$seurat_clusters)
colors <- pals::alphabet2(nclusters)
names(colors) <- NULL
```

## UMAP

If this was a bulk RNA-seq dataset, we would visualize the samples using principal components. However, doing this with single cell we get a visualization that is not getting advantage of the whole space, mostly because:

1. It is a linear transformation of the data
2. There are other components representing the data variation, so sometimes dots from a cluster overlay with dots from other cluster

```{r pc-analysis}
DimPlot(lung_data, reduction = "pca", cols = colors)
```

Therefore we need to depends on other dimensionality reduction methods like t-SNE, UMAP, etc. For example, we use UMAP and we can notice:

1. We can see that all clusters are represented, i.e. all are in front
2. The cells in each cluster are grouped together, e.g. for clusters 7 and 8 the difference is the most striking
3. The clusters that were close to each other are still close, e.g. `1 ~ 5 ~ 3`, `0 ~ 2`, etc
4. The clustering algorithm is not perfect, it looks like there are clusters that can be further separated, thus we need to increase the resolution

```{r umap-analysis}
lung_data <- RunUMAP(lung_data, dims = 1:10)
DimPlot(lung_data, reduction = "umap", cols = colors)
```

**Disclaimer:** UMAP is similar to t-SNE, and we would use the `runTSNE` instead of `runUMAP`:

```{r tsne-analysis, eval = FALSE}
lung_data <- RunTSNE(lung_data, dims = 1:10)
DimPlot(lung_data, reduction = "tsne", cols = colors)
```

## Find markers per cluster



```{r get-representative-genes}
lung_data_markers <- FindAllMarkers(
    lung_data, only.pos = TRUE,
    min.pct = 0.25, logfc.threshold = 0.25)
top_markers <- lung_data_markers %>% 
    group_by(cluster) %>% 
    top_n(n = 15, wt = avg_log2FC)
```



```{r representative-genes, out.width="90%", fig.width=7, fig.height=9}
DoHeatmap(lung_data, features = top_markers$gene,
          group.colors = colors, size = 3) +
  NoLegend()
```

## Final thoughts


