<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to single cell RNA-seq analysis • cisr2021singlecellRNA</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to single cell RNA-seq analysis">
<meta property="og:description" content="cisr2021singlecellRNA">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">cisr2021singlecellRNA</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/intro_sc_analysis.html">Workshop</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/welch16/cisr2021singlecellRNA">
    <span class="fas fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="intro_sc_analysis_files/header-attrs-2.7/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Introduction to single cell RNA-seq analysis</h1>
                        <h4 class="author">Rene Welch</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/welch16/cisr2021singlecellRNA/blob/master/vignettes/intro_sc_analysis.Rmd"><code>vignettes/intro_sc_analysis.Rmd</code></a></small>
      <div class="hidden name"><code>intro_sc_analysis.Rmd</code></div>

    </div>

    
    
<div id="workshop-description" class="section level2">
<h2 class="hasAnchor">
<a href="#workshop-description" class="anchor"></a>Workshop description</h2>
<p>In exploring and analyzing single cell RNA sequencing data, there are a number of key concepts, such as filtering, scaling, dimensionality reduction, hypothesis testing, clustering and visualization, that need to be understood.</p>
<p>The idea behind this small session is to provide tools to start analyzing a single cell RNA-seq dataset, we are going to focus on how to utilize the R package Seurat to provide a dimensionality reduction visualization. These notes are based on this <a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html">tutorial</a></p>
<div id="r-bioconductor-package-used" class="section level3">
<h3 class="hasAnchor">
<a href="#r-bioconductor-package-used" class="anchor"></a>R / Bioconductor package used</h3>
<ul>
<li><code>scRNAseq</code></li>
<li><code>Seurat</code></li>
</ul>
</div>
</div>
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>There are many steps involved in analyzing a single cell RNA-seq experiment. Usually, we start by aligning the sequence reads into a reference genome to quantify the number of reads mapped to each gene. This results into a table of counts, which we are going to use to perform statistical analysis using R. For single cell RNA-seq data, there are many tools to perform this step, but today we are going to utilize from an already curated dataset, and show how to process it with the R package <code>Seurat</code>.</p>
<p>First, let’s load all the packages we will need to analyze the data:</p>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="https://satijalab.org/seurat">Seurat</a></span>) <span class="co">## main package that we are going to use</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">SingleCellExperiment</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">scRNAseq</span>) 
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="http://tidyverse.tidyverse.org">tidyverse</a></span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="https://wilkelab.org/cowplot">cowplot</a></span>)
</pre></div>
</div>
<div id="getting-the-data" class="section level2">
<h2 class="hasAnchor">
<a href="#getting-the-data" class="anchor"></a>Getting the data</h2>
<p>First, we need to download a dataset. Fortunately, the <code>scRNAseq</code> package provides a variety of already curated datasets:</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="kw">lung_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scRNAseq/man/ZilionisLungData.html">ZilionisLungData</a></span>(<span class="st">"mouse"</span>)
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="kw">lung_data</span>) <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/make.unique.html">make.unique</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="kw">lung_data</span>), sep = <span class="st">"_"</span>)
<span class="fu">colData</span>(<span class="kw">lung_data</span>)[[<span class="st">"Barcode"</span>]] <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="kw">lung_data</span>)
<span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html">logcounts</a></span>(<span class="kw">lung_data</span>) <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span>(<span class="fl">1</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html">counts</a></span>(<span class="kw">lung_data</span>)))
<span class="kw">lung_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/as.Seurat.html">as.Seurat</a></span>(<span class="kw">lung_data</span>, project = <span class="st">"lung"</span>)
</pre></div>
<p>Now, Seurat provides all sorts of useful information on our dataset:</p>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="kw">lung_data</span>
<span class="co">#&gt; An object of class Seurat </span>
<span class="co">#&gt; 28205 features across 17549 samples within 1 assay </span>
<span class="co">#&gt; Active assay: RNA (28205 features, 0 variable features)</span>
<span class="co">#&gt;  1 dimensional reduction calculated: SPRING</span>
</pre></div>
<p>To access the metadata, i.e. information for each cell / droplet, we do:</p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="kw">lung_data</span><span class="op">@</span>meta.data)
<span class="co">#&gt;        Library Barcode Animal Run  Tissue Used Library.prep.batch Total.counts</span>
<span class="co">#&gt; bc0001   h_1_1  bc0001    h_1   1 healthy TRUE    round1_20151128        11915</span>
<span class="co">#&gt; bc0002   h_1_1  bc0002    h_1   1 healthy TRUE    round1_20151128        11675</span>
<span class="co">#&gt; bc0003   h_1_1  bc0003    h_1   1 healthy TRUE    round1_20151128        12536</span>
<span class="co">#&gt; bc0004   h_1_1  bc0004    h_1   1 healthy TRUE    round1_20151128        12499</span>
<span class="co">#&gt; bc0005   h_1_1  bc0005    h_1   1 healthy TRUE    round1_20151128         9586</span>
<span class="co">#&gt; bc0006   h_1_1  bc0006    h_1   1 healthy TRUE    round1_20151128        10553</span>
<span class="co">#&gt;        Percent.counts.from.mitochondrial.genes Most.likely.Immgen.cell.type</span>
<span class="co">#&gt; bc0001                                   0.210                 GN_Arth_SynF</span>
<span class="co">#&gt; bc0002                                   0.240                 GN_Arth_SynF</span>
<span class="co">#&gt; bc0003                                   2.728                  MF_PPAR-_Lu</span>
<span class="co">#&gt; bc0004                                   2.728                        MF_Lu</span>
<span class="co">#&gt; bc0005                                   2.378                        MF_Lu</span>
<span class="co">#&gt; bc0006                                   1.516                        Mo_Lu</span>
<span class="co">#&gt;        Major.cell.type Minor.subset</span>
<span class="co">#&gt; bc0001     Neutrophils           N1</span>
<span class="co">#&gt; bc0002     Neutrophils           N1</span>
<span class="co">#&gt; bc0003         MoMacDC         Mac4</span>
<span class="co">#&gt; bc0004         MoMacDC         Mac4</span>
<span class="co">#&gt; bc0005         MoMacDC         Mac4</span>
<span class="co">#&gt; bc0006         MoMacDC        Mono3</span>
</pre></div>
<p>This is a curated dataset, thus it already contains the cell labels. At the moment we are going to assume that we don’t know that yet.</p>
<p><strong>Quick note:</strong> This data could be a bit large for some laptops, so in case of trouble running this analysis, then it could be useful to pick another dataset. To list the datasets provided by the <code>scRNAseq</code> package use:</p>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/pkg/scRNAseq/man/listDatasets.html">listDatasets</a></span>()
</pre></div>
</div>
<div id="quality-control" class="section level2">
<h2 class="hasAnchor">
<a href="#quality-control" class="anchor"></a>Quality control</h2>
<p>Usually to do a quick quality control, the idea behind this step is that we would like to not have cells in our dataset with an abnormally large / small # of genes (we quantify the # of genes for a cell/droplet as the # of genes with at least 1 read). The reasons for them are:</p>
<ul>
<li>Cell doublets or multiplets may exhibit and abnormally high number of genes</li>
<li>Low-quality cells or empty droplets will often have few genes</li>
</ul>
<p>First, we add a pair of QC metric to the <code>meta.data</code>, and the explore their distribution:</p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="kw">lung_data</span>[[<span class="st">"nCount_RNA"</span>]] <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span>(<span class="kw">lung_data</span>[[<span class="st">"RNA"</span>]]<span class="op">@</span>counts)
<span class="kw">lung_data</span>[[<span class="st">"nFeature_RNA"</span>]] <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span>(<span class="kw">lung_data</span>[[<span class="st">"RNA"</span>]]<span class="op">@</span>counts, <span class="fl">2</span>, <span class="fu">function</span>(<span class="kw">x</span>)<span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(<span class="kw">x</span> <span class="op">&gt;</span> <span class="fl">0</span>))
<span class="kw">lung_data</span>[[<span class="st">"mt_perc"</span>]] <span class="op">&lt;-</span> <span class="kw">Seurat</span>::<span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/PercentageFeatureSet.html">PercentageFeatureSet</a></span>(<span class="kw">lung_data</span>, <span class="st">"^mt-"</span>)
<span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/VlnPlot.html">VlnPlot</a></span>(<span class="kw">lung_data</span>, features = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"nFeature_RNA"</span>, <span class="st">"nCount_RNA"</span>, <span class="st">"mt_perc"</span>), ncol = <span class="fl">3</span>)
</pre></div>
<p><img src="intro_sc_analysis_files/figure-html/brief-qc-1.png" width="90%" style="display: block; margin: auto;"></p>
<p><strong>Disclaimer</strong> It is always good to check the gene naming conventions. In the <code>Seurat</code> tutorial, the mitochondrial genes are searched by looking for the <code>MT</code> pattern at the start of every gene symbol. In this dataset, the pattern that we searched was <code>mt</code> so defining the percentage of mitochondrial genes in a cell with the <code>MT</code> would result in 0% for every cell!</p>
<p>Then, we compare the first two quantities, as expected both metrics are very correlated:</p>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FeatureScatter.html">FeatureScatter</a></span>(<span class="kw">lung_data</span>, feature1 = <span class="st">"nCount_RNA"</span>, feature2 = <span class="st">"nFeature_RNA"</span>,
               cols = <span class="st">"black"</span>) <span class="op">+</span>
    <span class="fu">geom_smooth</span>(se = <span class="fl">FALSE</span>) <span class="op">+</span>
    <span class="fu">geom_vline</span>(xintercept = <span class="fl">12e3</span>, linetype = <span class="fl">2</span>, colour = <span class="st">"red"</span>) <span class="op">+</span>
    <span class="fu">geom_hline</span>(yintercept = <span class="fl">3e3</span>, linetype = <span class="fl">2</span>, colour = <span class="st">"red"</span>)
</pre></div>
<p><img src="intro_sc_analysis_files/figure-html/metrics-corr-1.png" width="80%" style="display: block; margin: auto;"></p>
<p>We remove the cells with more than 12K reads or more than 3K genes and normalize the data, the default method uses logarithms with a <em>pseudo-count</em> and a scaling factor, which could be good for a first pass. However, many more methods for normalization have been developed, thus it may is possible that there are methods more useful for different datasets.</p>
<div class="sourceCode" id="cb8"><pre class="downlit">
<span class="kw">lung_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span>(<span class="kw">lung_data</span>,
                    subset = <span class="kw">nFeature_RNA</span> <span class="op">&lt;=</span> <span class="fl">3e3</span> <span class="op">&amp;</span> <span class="kw">nFeature_RNA</span> <span class="op">&lt;=</span> <span class="fl">12e3</span> <span class="op">&amp;</span>
                      <span class="kw">mt_perc</span> <span class="op">&lt;=</span> <span class="fl">20</span> <span class="op">&amp;</span> <span class="kw">Used</span>)
<span class="co">## writing the default parameters, explicitly</span>
<span class="kw">lung_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/NormalizeData.html">NormalizeData</a></span>(<span class="kw">lung_data</span>,
                           normalization.method = <span class="st">"LogNormalize"</span>,
                           scale.factor = <span class="fl">10000</span>)
</pre></div>
</div>
<div id="identification-of-most-variable-genes" class="section level2">
<h2 class="hasAnchor">
<a href="#identification-of-most-variable-genes" class="anchor"></a>Identification of most variable genes</h2>
<p>The figure shows some association between the average and variances gene expression. The point at the end is that focusing in the most variable genes helps to highlight the signal of single cell datasets. We are going to identify the top 2K most variable genes, but lets keep in mind that this may remove some genes of interest.</p>
<div class="sourceCode" id="cb9"><pre class="downlit">
<span class="kw">lung_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindVariableFeatures.html">FindVariableFeatures</a></span>(<span class="kw">lung_data</span>, selection.method = <span class="st">"vst"</span>,
                                  nfeatures = <span class="fl">2000</span>)
<span class="kw">plot1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/VariableFeaturePlot.html">VariableFeaturePlot</a></span>(<span class="kw">lung_data</span>, log = <span class="fl">TRUE</span>)
<span class="kw">top_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/VariableFeatures.html">VariableFeatures</a></span>(<span class="kw">lung_data</span>), <span class="fl">20</span>)
<span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/LabelPoints.html">LabelPoints</a></span>(<span class="kw">plot1</span>, points = <span class="kw">top_genes</span>, repel = <span class="fl">TRUE</span>)
</pre></div>
<p><img src="intro_sc_analysis_files/figure-html/variable-features-1.png" width="80%" style="display: block; margin: auto;"></p>
</div>
<div id="principal-components-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#principal-components-analysis" class="anchor"></a>Principal Components Analysis</h2>
<p>The next steps are to scale the data and to compute the principal components of the data. This is necessary because <code>Seurat</code> will utilize the components to calculate how do cells relate to each other. These steps are slow when we use all the genes in the data, so it is common to use the list of most variable genes.</p>
<div class="sourceCode" id="cb10"><pre class="downlit">
<span class="kw">variable_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/VariableFeatures.html">VariableFeatures</a></span>(object = <span class="kw">lung_data</span>)
<span class="kw">lung_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/ScaleData.html">ScaleData</a></span>(<span class="kw">lung_data</span>, features = <span class="kw">variable_genes</span>)
<span class="kw">lung_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/RunPCA.html">RunPCA</a></span>(<span class="kw">lung_data</span>, features = <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/VariableFeatures.html">VariableFeatures</a></span>(object = <span class="kw">lung_data</span>))
</pre></div>
<p>For example, we can visualize top genes that are driving the variance explained by each component:</p>
<div class="sourceCode" id="cb11"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/VizDimLoadings.html">VizDimLoadings</a></span>(<span class="kw">lung_data</span>, dims = <span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, reduction = <span class="st">"pca"</span>, ncol = <span class="fl">3</span>)
</pre></div>
<p><img src="intro_sc_analysis_files/figure-html/viz-loadings-1.png" width="90%" style="display: block; margin: auto;"></p>
<p>Then, the next question to solve is how many components to pick. There are many alternatives to solve this question, but the simplest alternative is to pick them by the Elbow criteria, which is to compare the std. deviation explained by each components with its rank, and pick the previous components before we observe and elbow pattern.</p>
<div class="sourceCode" id="cb12"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/ElbowPlot.html">ElbowPlot</a></span>(<span class="kw">lung_data</span>)
</pre></div>
<p><img src="intro_sc_analysis_files/figure-html/pick-pc-components-1.png" width="700" style="display: block; margin: auto;"></p>
<p>This is not an exact approach, so we according to the figure above, we would pick the top 2, 7 or 12 components.</p>
</div>
<div id="clustering-the-cells" class="section level2">
<h2 class="hasAnchor">
<a href="#clustering-the-cells" class="anchor"></a>Clustering the cells</h2>
<p>This is the most important part of the analysis so far, <code>Seurat</code> applies a graph based clustering approach, and to determine the number of clusters it is going to be necessary to tune the two parameters: <strong>the top # of principal components</strong> and the <strong>resolution</strong>.</p>
<ul>
<li>The <strong>top # principal components</strong> determines how detailed is going to be our approximation when computing the distance between any two cells</li>
<li>The <strong>resolution</strong> determines the size of the clusters that the algorithm finds. Smaller values will result in larger clusters or less clusters.</li>
</ul>
<p><code>Seurat</code> utilizes the Louvain algorithm to cluster the cells in a neighbor graph, and it works by maximizing the “modularity” of the cells network, so it tries to group together the nodes that very similar among them, and to maximize the distance between different clusters / modules.</p>
<div class="sourceCode" id="cb13"><pre class="downlit">
<span class="kw">lung_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindNeighbors.html">FindNeighbors</a></span>(<span class="kw">lung_data</span>, dims = <span class="fl">1</span><span class="op">:</span><span class="fl">7</span>)
<span class="kw">lung_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindClusters.html">FindClusters</a></span>(<span class="kw">lung_data</span>, resolution = <span class="fl">.3</span>)
<span class="co">#&gt; Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Number of nodes: 15904</span>
<span class="co">#&gt; Number of edges: 498213</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Running Louvain algorithm...</span>
<span class="co">#&gt; Maximum modularity in 10 random starts: 0.9347</span>
<span class="co">#&gt; Number of communities: 8</span>
<span class="co">#&gt; Elapsed time: 2 seconds</span>
<span class="co">## this is to make the cluster colors agree below</span>
<span class="kw">nclusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nlevels.html">nlevels</a></span>(<span class="kw">lung_data</span><span class="op">@</span>meta.data<span class="op">$</span><span class="kw">seurat_clusters</span>)
<span class="kw">colors</span> <span class="op">&lt;-</span> <span class="kw">pals</span>::<span class="fu"><a href="https://rdrr.io/pkg/pals/man/discrete.html">alphabet2</a></span>(<span class="kw">nclusters</span>)
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="kw">colors</span>) <span class="op">&lt;-</span> <span class="kw">NULL</span>
</pre></div>
</div>
<div id="umap" class="section level2">
<h2 class="hasAnchor">
<a href="#umap" class="anchor"></a>UMAP</h2>
<p>To get a UMAP representation of the data we utilize:</p>
<div class="sourceCode" id="cb14"><pre class="downlit">
<span class="kw">lung_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/RunUMAP.html">RunUMAP</a></span>(<span class="kw">lung_data</span>, dims = <span class="fl">1</span><span class="op">:</span><span class="fl">7</span>)
</pre></div>
<p>Then we can visualize how the cells are distributed under different discrete <code>meta.data</code> variables. The important thing to notice in these figures is that the cells are mixed regardless of batch / tissue.</p>
<div class="sourceCode" id="cb15"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/pkg/cowplot/man/plot_grid.html">plot_grid</a></span>(
  <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/DimPlot.html">DimPlot</a></span>(<span class="kw">lung_data</span>, reduction = <span class="st">"umap"</span>, group.by = <span class="st">"Library.prep.batch"</span>),
  <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/DimPlot.html">DimPlot</a></span>(<span class="kw">lung_data</span>, reduction = <span class="st">"umap"</span>, group.by = <span class="st">"Tissue"</span>), nrow = <span class="fl">1</span>
)
</pre></div>
<p><img src="intro_sc_analysis_files/figure-html/umap-extra-1.png" width="100%" style="display: block; margin: auto;"></p>
<p>If this was a bulk RNA-seq dataset, we would visualize the samples using principal components. However, doing this with single cell we get a visualization that is not getting advantage of the whole space, mostly because:</p>
<ol style="list-style-type: decimal">
<li>It is a linear transformation of the data</li>
<li>There are other components representing the data variation, so sometimes dots from a cluster overlay with dots from other cluster</li>
</ol>
<div class="sourceCode" id="cb16"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/DimPlot.html">DimPlot</a></span>(<span class="kw">lung_data</span>, reduction = <span class="st">"pca"</span>, cols = <span class="kw">colors</span>)
</pre></div>
<p><img src="intro_sc_analysis_files/figure-html/pc-analysis-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Therefore we need to depends on other dimensionality reduction methods like t-SNE, UMAP, etc. For example, we use UMAP and we can notice:</p>
<ol style="list-style-type: decimal">
<li>We can see that all clusters are represented, i.e. all are in front</li>
<li>The cells in each cluster are grouped together, e.g. for clusters 7 and 8 the difference is the most striking</li>
<li>The clusters that were close to each other are still close, e.g. <code>1 ~ 3 ~ 7</code>, <code>0 ~ 2</code>, etc</li>
<li>The clustering algorithm is not perfect, it looks like there are clusters that can be further separated, thus we need to increase the resolution</li>
</ol>
<div class="sourceCode" id="cb17"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/DimPlot.html">DimPlot</a></span>(<span class="kw">lung_data</span>, reduction = <span class="st">"umap"</span>, cols = <span class="kw">colors</span>)
</pre></div>
<p><img src="intro_sc_analysis_files/figure-html/umap-analysis-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Finally, the table below illustrates that the clusters computed by <code>Seurat</code> approximately coincides with the major cell types found in the manuscript:</p>
<div class="sourceCode" id="cb18"><pre class="downlit">
<span class="kw">lung_data</span><span class="op">@</span>meta.data <span class="op">%&gt;%</span>
  <span class="fu">select</span>(<span class="kw">Major.cell.type</span>, <span class="kw">seurat_clusters</span>) <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>()
<span class="co">#&gt;                seurat_clusters</span>
<span class="co">#&gt; Major.cell.type    0    1    2    3    4    5    6    7</span>
<span class="co">#&gt;     B cells        1 2795    0    7    2    4    2    2</span>
<span class="co">#&gt;     Basophils      3    0   27    1    2    0    1    0</span>
<span class="co">#&gt;     MoMacDC       13    8   10    4 1081  616  637    0</span>
<span class="co">#&gt;     Neutrophils 5523    9 2438    7    8    1   31    0</span>
<span class="co">#&gt;     NK cells       0    1    0   90    0    0    0  539</span>
<span class="co">#&gt;     pDC            0    1    0    3    1   57    0    0</span>
<span class="co">#&gt;     T cells        2   35    2 1924    0    4    0   12</span>
</pre></div>
<p><strong>Disclaimer:</strong> UMAP is similar to t-SNE, and we would use the <code>runTSNE</code> instead of <code>runUMAP</code>:</p>
<div class="sourceCode" id="cb19"><pre class="downlit">
<span class="kw">lung_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/RunTSNE.html">RunTSNE</a></span>(<span class="kw">lung_data</span>, dims = <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>)
<span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/DimPlot.html">DimPlot</a></span>(<span class="kw">lung_data</span>, reduction = <span class="st">"tsne"</span>, cols = <span class="kw">colors</span>)
</pre></div>
</div>
<div id="find-markers-per-cluster" class="section level2">
<h2 class="hasAnchor">
<a href="#find-markers-per-cluster" class="anchor"></a>Find markers per cluster</h2>
<p>Finally to interpret the cluster, we can glance into few of the top genes that are representative into each cluster. This is done by using the <code>FindAllMarkers</code> function, which searches the differentially expressed genes when comparing the cells in every cluster against the others. Below we find all the markers and pick the top 10 genes in average log2 fold change, i.e. the genes with the highest relative expression in one cluster against the rest.</p>
<div class="sourceCode" id="cb20"><pre class="downlit">
<span class="kw">lung_data_markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindAllMarkers.html">FindAllMarkers</a></span>(
    <span class="kw">lung_data</span>, only.pos = <span class="fl">TRUE</span>,
    min.pct = <span class="fl">0.25</span>, logfc.threshold = <span class="fl">0.25</span>)
<span class="kw">top_markers</span> <span class="op">&lt;-</span> <span class="kw">lung_data_markers</span> <span class="op">%&gt;%</span> 
    <span class="fu">group_by</span>(<span class="kw">cluster</span>) <span class="op">%&gt;%</span> 
    <span class="fu">top_n</span>(n = <span class="fl">10</span>, wt = <span class="kw">avg_log2FC</span>)
</pre></div>
<p>We can visualize these markers into a heatmap, where the rows are a genes and each column is one of the many cells. The yellow / brown color mean that a gene is expressed in a given cell.</p>
<div class="sourceCode" id="cb21"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/DoHeatmap.html">DoHeatmap</a></span>(<span class="kw">lung_data</span>, features = <span class="kw">top_markers</span><span class="op">$</span><span class="kw">gene</span>,
          group.colors = <span class="kw">colors</span>, size = <span class="fl">3</span>) <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/SeuratTheme.html">NoLegend</a></span>()
</pre></div>
<p><img src="intro_sc_analysis_files/figure-html/representative-genes-1.png" width="90%" style="display: block; margin: auto;"></p>
<p>Changing the cluster names, we get:</p>
<div class="sourceCode" id="cb22"><pre class="downlit">
<span class="kw">lung_data</span><span class="op">@</span>meta.data<span class="op">$</span><span class="kw">seurat_clusters</span> <span class="op">&lt;-</span>
  <span class="kw">forcats</span>::<span class="fu"><a href="https://forcats.tidyverse.org/reference/fct_recode.html">fct_recode</a></span>(<span class="kw">lung_data</span><span class="op">@</span>meta.data<span class="op">$</span><span class="kw">seurat_clusters</span>,
    `Neutrophils` = <span class="st">"0"</span>,
    `B cells` = <span class="st">"1"</span>,
    `Basophils` = <span class="st">"2"</span>,
    `T cells` = <span class="st">"3"</span>,
    `MoMacDC1` = <span class="st">"4"</span>,
    `MoMacDC2` = <span class="st">"5"</span>,
    `MoMacDC3` = <span class="st">"6"</span>,
    `NK cells` = <span class="st">"7"</span>)
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="kw">colors</span>) <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span>(<span class="kw">lung_data</span><span class="op">@</span>meta.data<span class="op">$</span><span class="kw">seurat_clusters</span>)
<span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/DimPlot.html">DimPlot</a></span>(<span class="kw">lung_data</span>, reduction = <span class="st">"umap"</span>, cols = <span class="kw">colors</span>, group.by = <span class="st">"seurat_clusters"</span>, label = <span class="fl">TRUE</span>) <span class="op">+</span>
  <span class="fu">ggtitle</span>(<span class="kw">NULL</span>)
</pre></div>
<p><img src="intro_sc_analysis_files/figure-html/umap-analysis-truth-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Finally, we compare the composition of tumor vs healthy cells:</p>
<div class="sourceCode" id="cb23"><pre class="downlit">
<span class="kw">lung_data</span><span class="op">@</span>meta.data <span class="op">%&gt;%</span>
  <span class="fu">group_by</span>(<span class="kw">Tissue</span>) <span class="op">%&gt;%</span>
  <span class="kw">dplyr</span>::<span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html">count</a></span>(<span class="kw">seurat_clusters</span>) <span class="op">%&gt;%</span>
  <span class="fu">mutate</span>(p = <span class="kw">n</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(<span class="kw">n</span>)) <span class="op">%&gt;%</span>
  <span class="fu">ggplot</span>(<span class="fu">aes</span>(y = <span class="kw">Tissue</span>, x = <span class="kw">p</span>, fill = <span class="kw">seurat_clusters</span>)) <span class="op">+</span> <span class="fu">geom_col</span>() <span class="op">+</span>
  <span class="fu">theme_light</span>() <span class="op">+</span>
  <span class="fu">theme</span>(legend.position = <span class="st">"top"</span>, axis.title = <span class="fu">element_blank</span>(), axis.text = <span class="fu">element_text</span>(size = <span class="fl">14</span>),
        legend.text = <span class="fu">element_text</span>(size = <span class="fl">12</span>)) <span class="op">+</span>
  <span class="fu">scale_fill_manual</span>(values = <span class="kw">colors</span>) <span class="op">+</span>
  <span class="fu">scale_x_continuous</span>(labels = <span class="kw">scales</span>::<span class="fu"><a href="https://scales.r-lib.org//reference/label_percent.html">percent_format</a></span>(<span class="fl">1</span>), breaks = <span class="kw">scales</span>::<span class="fu"><a href="https://scales.r-lib.org//reference/breaks_pretty.html">breaks_pretty</a></span>(n = <span class="fl">4</span>))
</pre></div>
<p><img src="intro_sc_analysis_files/figure-html/cell-composition-1.png" width="100%" style="display: block; margin: auto;"></p>
</div>
<div id="final-thoughts" class="section level2">
<h2 class="hasAnchor">
<a href="#final-thoughts" class="anchor"></a>Final thoughts</h2>
<p>There is no an infallible recipe for analyzing any dataset, in particular single cell experiments had become more complex. In this document, we show only one strategy to start analyzing these sort of experiments.</p>
<p>In this tutorial, we utilized data of a recent experiment and showed with simple instructions how to approximately replicate some of the clustering.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Rene Welch.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
